RATIS-1990. Refactor appendEntries processing to support reference count

duongkame opened the pull request:
## What changes were proposed in this pull request?

Refactor appendEntries processing to support reference count

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/RATIS-1990

## How was this patch tested?

CI.

duongkame commented:
c.c @szetszwo

duongkame commented:
Moved back to draft to debug a relevant error. Somehow TestRaftWithSimulatedRpc (testWithLoad and testOldLeaderNotCommit) creates a situation which two log entries (different terms) come to a server for the same index. The later fails because the TransactionContext has already already created and cached for the first one. 
This seems to be always the case, but now starts to be exposed because now we apply referent count for appendEntries. 
@szetszwo 

```
Caused by: java.lang.IllegalStateException: entry.term: expected == 4 but computed == 5
	at org.apache.ratis.util.Preconditions.assertTrue(Preconditions.java:76)
	at org.apache.ratis.util.Preconditions.assertSame(Preconditions.java:81)
	at org.apache.ratis.statemachine.TransactionContext.wrap(TransactionContext.java:104)
	at org.apache.ratis.statemachine.impl.TransactionContextImpl.wrap(TransactionContextImpl.java:136)
	at org.apache.ratis.server.raftlog.segmented.SegmentedRaftLog.appendEntryImpl(SegmentedRaftLog.java:434)
	at org.apache.ratis.server.raftlog.RaftLogBase.lambda$appendEntry$11(RaftLogBase.java:351)
	at org.apache.ratis.server.raftlog.RaftLogSequentialOps$Runner.runSequentially(RaftLogSequentialOps.java:78)
	at org.apache.ratis.server.raftlog.RaftLogBase.appendEntry(RaftLogBase.java:351)
	at org.apache.ratis.server.raftlog.segmented.SegmentedRaftLog.appendImpl(SegmentedRaftLog.java:477)
	at org.apache.ratis.server.raftlog.RaftLogBase.lambda$append$12(RaftLogBase.java:358)
	at org.apache.ratis.server.raftlog.RaftLogSequentialOps$Runner.runSequentially(RaftLogSequentialOps.java:69)
	at org.apache.ratis.server.raftlog.RaftLogBase.append(RaftLogBase.java:358)
	at org.apache.ratis.server.impl.RaftServerImpl.appendEntriesAsync(RaftServerImpl.java:1653)
	at org.apache.ratis.server.impl.RaftServerImpl.appendEntriesAsync(RaftServerImpl.java:1517)
	at org.apache.ratis.server.impl.RaftServerImpl.appendEntries(RaftServerImpl.java:1500)
	at org.apache.ratis.server.impl.RaftServerProxy.appendEntries(RaftServerProxy.java:663)
	at org.apache.ratis.server.simulation.SimulatedServerRpc$1.handleRequest(SimulatedServerRpc.java:163)
```

szetszwo commented:
> ... TestRaftWithSimulatedRpc (testWithLoad and testOldLeaderNotCommit) creates a situation ...

Could we fix `TestRaftWithSimulatedRpc`?

duongkame commented:
> Could we fix `TestRaftWithSimulatedRpc`?

I'm working on it, will republish this PR after fixing it.

duongkame commented:
@szetszwo  Regarding `TestRaftWithSimulatedRpc`. What happened is:
1. `s2` was the leader of term `2`.
2. `s2` received a client request, added to log as `(t:2, i:1475)`.
3. `s0` received `appendEntries` from `s2` for `(t:2, i:1475)`. A `TransactionContext` is [created](https://github.com/apache/ratis/blob/master/ratis-server/src/main/java/org/apache/ratis/server/raftlog/segmented/SegmentedRaftLog.java#L477) for `(t:2, i:1475)` during `appendEntries` (I think this is wrong). 
4. `1475` hadn't been replcated to majority of nodes,  `s0` hadn't applied it yet.
5. Leader changed to `s1`. (`s1` hadn't received `appendEntries` for `(t:2, i:1475)`).
6. `s1` received a client request, appended it as `1475`
7. `s1` decreased `s0`'s `nextIndex`: `s1@group-9EAE5FF30820->s0: decreaseNextIndex nextIndex: updateUnconditionally 1475 -> 1474`
8. `s0` received `appendEntries` from `s2` for `(t:2, i:1475)`, it looked up `TransactionContext` and saw one created per 3. Assertion `Preconditions.assertSame(getLogEntry().getTerm(), entry.getTerm(), "entry.term");` fails.

I believe the problem is from the fact that TransactionContext is [created](https://github.com/apache/ratis/blob/master/ratis-server/src/main/java/org/apache/ratis/server/raftlog/segmented/SegmentedRaftLog.java#L477) during `appendEntries`.

``` 
    // in SegmentedRaftLog#appendImpl
     for (int i = index; i < entries.size(); i++) {
        final LogEntryProto entry = entries.get(i);
        futures.add(appendEntry(entry, server.getTransactionContext(entry, true)));
      }
```
`appendEntries` should lookup `TransactionContext` only, `TransactionContext` should be created once the entry is applied.
``` 
        futures.add(appendEntry(entry, server.getTransactionContext(entry, false)));
```

Look at the failing test per https://github.com/apache/ratis/actions/runs/7516329635/artifacts/1167720078, `org.apache.ratis.server.simulation.TestRaftWithSimulatedRpc-output.txt`.

szetszwo commented:
> Regarding TestRaftWithSimulatedRpc. What happened is: ...

@duongkame , Thanks a lot for digging it up!  Could you file a bug?  Please assign it to me (unless you like to work on it.)

duongkame commented:
> Could you file a bug? Please assign it to me (unless you like to work on it.)
Sure, filed [RATIS-2001](https://issues.apache.org/jira/browse/RATIS-2001) and assigned it to you. @szetszwo 

I temporarily removed the same term assertion to unblock this PR.

szetszwo commented on ratis-server-api/src/main/java/org/apache/ratis/server/raftlog/RaftLogSequentialOps.java:
```diff
@@ -132,21 +133,34 @@ default CompletableFuture<Long> appendEntry(LogEntryProto entry, TransactionCont
+   * For each entry, implementations of this method should retain the counter, process it and then release.
    */
-  List<CompletableFuture<Long>> append(List<LogEntryProto> entries);
+  default List<CompletableFuture<Long>> append(List<LogEntryProto> entries, ReferenceCountedObject<?> entriesRef) {
```
This should be
```java
  default List<CompletableFuture<Long>> append(ReferenceCountedObject<List<LogEntryProto>> entriesRef) {
    try(UncheckedAutoCloseableSupplier<List<LogEntryProto>> entries = entriesRef.retainAndReleaseOnClose()) {
      return append(entries.get());
    }
  }
```

szetszwo commented on ratis-server-api/src/main/java/org/apache/ratis/server/raftlog/RaftLogSequentialOps.java:
```diff
@@ -132,21 +133,34 @@ default CompletableFuture<Long> appendEntry(LogEntryProto entry, TransactionCont
+   * @deprecated use {@link #append(List, ReferenceCountedObject)}.
+   */
+  @Deprecated
+  default List<CompletableFuture<Long>> append(List<LogEntryProto> entries) {
+    return append(entries, null);
+  }
```
This should throw `UnsupportedOperationException`.

szetszwo commented on ratis-server/src/main/java/org/apache/ratis/server/raftlog/memory/MemoryRaftLog.java:
```diff
@@ -181,7 +192,8 @@ public long getStartIndex() {
 
   @Override
-  public List<CompletableFuture<Long>> appendImpl(List<LogEntryProto> logEntryProtos) {
+  public List<CompletableFuture<Long>> appendImpl(List<LogEntryProto> logEntryProtos,
+      ReferenceCountedObject<?> entriesRef) {
```
It should be
```java
  public List<CompletableFuture<Long>> appendImpl(ReferenceCountedObject<List<LogEntryProto>> entriesRef) {
```

szetszwo commented on ratis-server/src/main/java/org/apache/ratis/server/raftlog/memory/MemoryRaftLog.java:
```diff
@@ -212,9 +224,13 @@ public List<CompletableFuture<Long>> appendImpl(List<LogEntryProto> logEntryProt
         futures = new ArrayList<>(logEntryProtos.size() - index);
       }
+
+      Function<LogEntryProto, ReferenceCountedObject<LogEntryProto>> wrap = entriesRef != null ?
+          entriesRef::delegate : ReferenceCountedObject::wrap;
       for (int i = index; i < logEntryProtos.size(); i++) {
         LogEntryProto logEntryProto = logEntryProtos.get(i);
-        this.entries.add(logEntryProto);
+        ReferenceCountedObject<LogEntryProto> entryRef = wrap.apply(logEntryProto);
+        this.entries.add(entryRef);
         futures.add(CompletableFuture.completedFuture(logEntryProto.getIndex()));
       }
```
Then, we could use `entriesRef` to delegate:
```java
      for (int i = index; i < logEntryProtos.size(); i++) {
        LogEntryProto logEntryProto = logEntryProtos.get(i);
        entries.add(entriesRef.delegate(logEntryProto));
        futures.add(CompletableFuture.completedFuture(logEntryProto.getIndex()));
      }
```

szetszwo reviewed (commented):
@duongkame , thanks a lot for working on this complicated issue!  Please see the comments inlined and also https://issues.apache.org/jira/secure/attachment/13066076/1011_review.patch

szetszwo commented:
> ... Sure, filed RATIS-2001 ...

RATIS-2001 is merged.  Please update this with master.

duongkame commented:
> [RATIS-2001](https://issues.apache.org/jira/browse/RATIS-2001) is merged. Please update this with master.

Thanks a lot for the quick fix and the detailed review. @szetszwo 
I've updated the code with the latest master and addressed your comments.

szetszwo reviewed (approved):
+1 the change looks good.

duongkame commented:
Thanks for the review @szetszwo
